src/inmanta/parser/plyInmantaParser.py

```diff
@@ -731,40 +733,52 @@ def p_constant_rstring(p: YaccProduction) -> None:
 
 def p_constant_mls(p: YaccProduction) -> None:
     "constant : mls"
-    p[0] = get_string_ast_node(p[1], p[1].location)
+    p[0] = get_string_ast_node(p[1], True)
     attach_from_string(p)
 
 
 format_regex = r"""({{\s*([\.A-Za-z0-9_-]+)\s*}})"""
 format_regex_compiled = re.compile(format_regex, re.MULTILINE | re.DOTALL)
 
 
-def get_string_ast_node(string: LocatableString, location: Location) -> Union[Literal, StringFormat]:
-    match_obj = format_regex_compiled.findall(str(string))
-    if len(match_obj) == 0:
+def get_string_ast_node(string: LocatableString, mls: boolean) -> Union[Literal, StringFormat]:
+    matches = [[str(string)[m.start() : m.end()], m.start(), m.end()] for m in format_regex_compiled.finditer(str(string))]
+    if len(matches) == 0:
         return Literal(str(string))
-    return create_string_format(string, match_obj, location)
 
-
-def create_string_format(format_string: LocatableString, variables: List[List[str]], location: Location) -> StringFormat:
+    offset = 3 if mls else 1  # len(""")  or len(') or len(")
+    init: List[str]
+    match: List[str]
+    locatable_matches: List(LocatableString) = []
+    for obj in matches:
+        init, match, _ = (str(string)[start:end].splitlines() for start, end in pairwise((0, obj[1], obj[2], len(obj))))
+        start_line: int = string.lnr + len(init) - 1
+        end_line: int = start_line + len(match) - 1
+        len_last_init_line: int = len(init[-1]) if init else 0
+        start_char: int = len(init[-1]) + offset + string.start if len(init) == 1 else len_last_init_line + 1
+        end_char: int = len(match[-1]) + start_char - 1 if len(match) == 1 else len(match[-1])
+        range: Range = Range(string.location.file, start_line, start_char, end_line, end_char)
+        locatable_string = LocatableString("".join(match), range, string.lexpos, string.namespace)
+        locatable_matches.append(locatable_string)
+    return create_string_format(string, locatable_matches)
+
+
+def create_string_format(format_string: LocatableString, variables: List[LocatableString]) -> StringFormat:
     """
     Create a string interpolation statement
     """
     _vars = []
-
-    for var_str in variables:
-        var_parts = var_str[1].split(".")
-        ref = Reference(var_parts[0])
-        ref.namespace = namespace
-
+    for var in variables:
+        var_parts: List[str] = str(var)[2:-2].strip().split(".")
+        ref_locatable_string = LocatableString(var_parts[0], var.location, var.lexpos, var.namespace)
```

Comment:
I see what you meant now. I would prefer this (and the `AttributeReference` one) to be accurate but I don't think it's worth too much effort. That said, I think getting the correct range should be quite a lot simpler here than for the loop above. Something like the snippet below could get you the required offset for each attribute:
```python
attribute_offsets: Iterator[int] = accumulate(var_parts[1:], lambda acc, part: acc + len(part) + 1, initial=len(var_parts[0] + 1))
for attr, offset in zip(var_parts[1:], attribute_offsets):
    ...
```

If you discover it turns out to be more complex, I'm fine with keeping this as is for now.

Reply:
I agree to that